download/src/lib.rs

```diff
@@ -61,74 +61,81 @@ pub fn download_to_path_with_backend(
+                        let mut downloaded_so_far = 0;
+                        loop {
+                            let n = partial.read(&mut buf)?;
+                            downloaded_so_far += n as u64;
+                            if n == 0 {
+                                break;
+                            }
+                            cb(Event::DownloadDataReceived(&buf[..n]))?;
                         }
-                        cb(Event::DownloadDataReceived(&buf[..n]))?;
-                    }
 
-                    downloaded_so_far
+                        downloaded_so_far
+                    } else {
+                        let file_info = partial.metadata()?;
+                        file_info.len()
+                    }
                 } else {
-                    let file_info = partial.metadata()?;
-                    file_info.len()
-                }
-            } else {
-                0
-            };
-
-            let mut possible_partial = OpenOptions::new()
-                .write(true)
-                .create(true)
-                .open(path)
-                .context("error opening file for download")?;
+                    0
+                };
 
-            possible_partial.seek(SeekFrom::End(0))?;
-
-            (possible_partial, downloaded_so_far)
-        } else {
-            (
-                OpenOptions::new()
+                // TODO: blocking call
+                let mut possible_partial = OpenOptions::new()
                     .write(true)
                     .create(true)
                     .open(path)
-                    .context("error creating file for download")?,
-                0,
-            )
-        };
+                    .context("error opening file for download")?;
 
-        let file = RefCell::new(file);
+                possible_partial.seek(SeekFrom::End(0))?;
 
-        download_with_backend(backend, url, resume_from, &|event| {
-            if let Event::DownloadDataReceived(data) = event {
-                file.borrow_mut()
-                    .write_all(data)
-                    .context("unable to write download to disk")?;
-            }
-            match callback {
-                Some(cb) => cb(event),
-                None => Ok(()),
-            }
-        })?;
+                (possible_partial, downloaded_so_far)
+            } else {
+                (
+                    OpenOptions::new()
+                        .write(true)
+                        .create(true)
+                        .open(path)
+                        .context("error creating file for download")?,
+                    0,
+                )
+            };
 
-        file.borrow_mut()
-            .sync_data()
-            .context("unable to sync download to disk")?;
+            let file = RefCell::new(file);
 
-        Ok(())
-    }()
+            // TODO: the sync callback will stall the async runtime if IO calls block, which is OS dependent. Rearrange.
```

Comment:
Should this be addressed before merging?

Reply:
No. Same as above, we should do so eventually.

The cb passed in does two things:
 - write a block of the file, which can/will block
 - calls the next layer out callback, which will be our download rate tracker

As we only download one file at the moment, there is no performance concern about the IO blocking - a rearranged async callback would want to prevent more data being downloaded to avoid OOM if disk IO were to stall (e.g. on NFS); thus this is appropriate for now, and when we make it async, it will be in order of supporting concurrent download of files, which is a different feature that we can all bring together